### DISCUSSION TITLE
Regarding parking env configuration

### DISCUSSION THREAD

**Discussion by HarshkumarBorad** [eleurent/highway-env#361] [Q&A] [answered]
1. Can I add boundaries to the parking area?
2. How to specify a location for other vehicles in parking area?

Thanks in advance.

**Comment by eleurent** [comment]
Hi,
In both cases, you have to modify the env's `_create_vehicles` method:

https://github.com/eleurent/highway-env/blob/742ce2e69bf00169f2b250e0388c6b9926c026ec/highway_env/envs/parking_env.py#L144

1. you can add boundaries by adding `Obstacle`s to the scene (`self.road.objects`) and make e.g. four walls (set the LENGTH/WIDTH/position/heading of each obstacle appropriately)
2. you can proceed similarly and add vehicles. this was already done in the Summon env, which is a fork of Parking with added parked vehicles: https://github.com/eleurent/highway-env/blob/742ce2e69bf00169f2b250e0388c6b9926c026ec/highway_env/envs/summon_env.py#L70

maybe this summon env should be made a configuration of the parking env (how many parked vehicles, presence of boundaries or not), instead of its own environment variant.

**Reply by HarshkumarBorad** [reply]
I will try to do so. Thank you so much for reply.

**Reply by HarshkumarBorad** [reply]
Hello there again,
I was able to modify and add boundaries. But, sometimes ego vehicles is colliding with it and sometimes not. I am attaching the file here. Could you please suggest to me a possible solution? 

Thanks and Regards,
Harshkumar Borad

https://user-images.githubusercontent.com/72328087/191064105-a63321ad-8c13-4e57-9e65-4cab41df77da.mp4

https://user-images.githubusercontent.com/72328087/191064111-f9749be4-2b71-4b77-a0f3-923ce06c69ed.mp4

**Reply by eleurent** [reply]
Mmmmh that is bad indeed, I'll have a look

**Reply by eleurent** [reply]
I think I found out why. If you look at the collision detection check, it starts with a coarse pre-check that is faster than the actual accuracte collision check:

https://github.com/eleurent/highway-env/blob/742ce2e69bf00169f2b250e0388c6b9926c026ec/highway_env/vehicle/objects.py#L103

This check verifies that the vehicle is within a circle centered on the obstacle's position within the obstacle's radius (called `diagonal `here), initialised to `np.sqrt(self.LENGTH**2 + self.WIDTH**2)` here:

https://github.com/eleurent/highway-env/blob/742ce2e69bf00169f2b250e0388c6b9926c026ec/highway_env/vehicle/objects.py#L49

So when you create the obstacle, if you change its length/width you should also update its diagonal:
```
        obstacle = Obstacle(self.road, [0, 15])
        obstacle.LENGTH, obstacle.WIDTH = (50, 2)
        obstacle.diagonal = np.sqrt(obstacle.LENGTH**2 + obstacle.WIDTH**2)
        self.road.objects.append(obstacle)
```

And I just realised that I am using only one of the two objects' radii to do the fast check, so this assumes that the two radii are the same which is not correct. I'll push a fix.

**Reply by eleurent** [reply]
I just pushed https://github.com/eleurent/highway-env/commit/89840aa9d599611cac4ccc0274cf16e270c5ee10, so you can just update the obstacle's diagonal as described above and that should do the trick :)

**Comment by HarshkumarBorad** [comment]
Hello, 

Thank you again for the last solution. It worked perfectly. Now I am back with another issue. Sometimes the goal is assigned to the parking spot which is already occupied by another car. Could you please suggest a possible solution for this? I will attach a few sample videos below. 

Thanks and Regards, 
Harhkumar Borad

https://user-images.githubusercontent.com/72328087/193405984-c54fce95-8348-4b67-b4dc-5986a84fe2f4.mp4
https://user-images.githubusercontent.com/72328087/193405952-360b9ad2-c37a-4ad0-92e3-f7dde9be5799.mp4

**Reply by eleurent** [reply]
You're right! Following this discussion, I pushed https://github.com/eleurent/highway-env/commit/d998fd857456c4e53262472f52f4fbae8371fd06 which does the following:

- Add support of adding vehicles and walls to ParkingEnv
- Fix the bug where a vehicle can spawn at the same location as the goal
- remove SummonEnv as it is now just a special case (configuration) of ParkingEnv

**Reply by HarshkumarBorad** [reply]
Hello there, 
Thank you for the response. Sorry to bother you again but, I am having the same issue as earlier for a few episodes. I will attach the file below. 

https://user-images.githubusercontent.com/72328087/193417742-c1b1366f-dcec-4bfe-a882-38d47599f96a.mp4

https://user-images.githubusercontent.com/72328087/193417754-de7b832d-036b-4447-9049-725482441fab.mp4

Thanks and Regards,
Harshkumar Borad

**Reply by eleurent** [reply]
Ok that's strange, this line was supposed to deal with these cases:
https://github.com/eleurent/highway-env/blob/954a0ab4a1ee9333c9f717e44c463592f79089ab/highway_env/envs/parking_env.py#L170

I'll investigate.

**Comment by cpatel245** [comment, marked as answer]
Hello @HarshkumarBorad 

https://github.com/eleurent/highway-env/blob/954a0ab4a1ee9333c9f717e44c463592f79089ab/highway_env/envs/parking_env.py#L168-L172

Try changing those lines in above snippet as shown in below and see it it works...

```
for v in self.road.vehicles:  # Prevent early collisions
    if v is not self.vehicle and (np.all(v.position == self.goal.position) or np.all(v.position == self.vehicle.position)):
        self.road.vehicles.remove(v)
```

**Reply by HarshkumarBorad** [reply]
Hello @crypto-g245,

Yes, it worked perfectly fine. Thank you so much for the suggestions. 

Thanks and Regards,
Harshkumar Borad

**Reply by eleurent** [reply]
Thanks @crypto-g245 for the suggestion, glad that it works!
I don't understand though, if the positions are equal (np.all(A == B)), then the current condition that checks norm(A-B)<20 should also be True, right?

**Reply by cpatel245** [reply]
Hello @eleurent , 

Yes, those conditions with norm is also somewhat right, what I mentioned in my previous reply was just a hotfix for the bug. 
for more general scenario, I agree, norm will be the right condition. 

While debugging I saw that increase (in x-coordinate) in distance with those adjacent parking spots are in multiple of 2. 

![image](https://user-images.githubusercontent.com/51653105/194817023-2319cb49-588f-4d94-bfbf-73999926a1a8.png)

it would be great if @HarshkumarBorad could test also this case with norm but with changed threshold. 
Please change threshold in below lines from 20 to 2 (as shown in code block below) and see if it provides the desired results 😊. 

https://github.com/eleurent/highway-env/blob/954a0ab4a1ee9333c9f717e44c463592f79089ab/highway_env/envs/parking_env.py#L168-L172

```
for v in self.road.vehicles:  # Prevent early collisions
    if v is not self.vehicle and (
            np.linalg.norm(v.position - self.goal.position) <= 2 or
            np.linalg.norm(v.position - self.vehicle.position) <= 2):
        self.road.vehicles.remove(v)
```

**Comment by HarshkumarBorad** [comment]
Hello there,
I would like to ask if anyone can suggest an algorithm for the below-attached scenarios. Currently, I am using SAC and TQC which were given in the examples. I have tried TD3 and DDPG but I am not able to get good results. 
1. Empty Parking Lots
![image_0](https://user-images.githubusercontent.com/72328087/198621653-c79944c7-6249-4b2c-943c-7158b2ab1b2b.png)

2. Parking Lots with parked cars
![image_2](https://user-images.githubusercontent.com/72328087/198620837-fe120ec1-d2e9-4dca-87c6-a709a3643b7f.png)

3. Parking Lots with parked cars and dynamic vehicles
![image_1](https://user-images.githubusercontent.com/72328087/198620765-491f38ee-31e5-4550-b294-31319436125f.png)


Thanks and Regards,
Harshkumar Borad

**Reply by eleurent** [reply]
Beside the algorithm, you may want to check that the observation you are using contains information about the locations of other vehicles in the scene (the default observation in the base parking env only contains agent and goal positions)

**Reply by HarshkumarBorad** [reply]
Hello, Thanks for the last suggestion. 

I am back again with one other question. I would like to plot a future trajectory as a waypoint or the line which agent is going to follow. Could you please suggest the possible solutions for that? 

Thanks and Regards,
Harshkumar Borad

**Reply by eleurent** [reply]
Sure, you can take a look at these function:

https://github.com/eleurent/highway-env/blob/11452929d3eb827ac6c4d547605d5eea19039429/highway_env/envs/common/graphics.py#L70

It allows you to provide a sequence of future actions, which is used to unroll the future trajectory of the vehicle, and this trajectory is then rendered below:

https://github.com/eleurent/highway-env/blob/11452929d3eb827ac6c4d547605d5eea19039429/highway_env/envs/common/graphics.py#L101

(you could edit this rendering to use waypoints / lines instead of the current implementation, which shows transparent vehicles as in this gif 
![](https://raw.githubusercontent.com/eleurent/highway-env/gh-media/docs/media/highway-env.gif?raw=true)
